[3D] qgs3dmapcanvaswidget: Prevent configure dialog from being modal

ptitjano opened the pull request:
When changing a parameter, it is difficult to check the result without
closing the parameter dialog because it always stays on top.
This issue is fixed by removing its modal state.

cc @wonder-sk @benoitdm-oslandia

ptitjano committed:
qgs3dmapcanvaswidget: Make configuration dialog non-modal

When changing a parameter, it is difficult to check the result without
closing the parameter dialog because it always stays on top.
This issue is fixed by removing its modal state.

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -387,15 +387,16 @@ void Qgs3DMapCanvasWidget::resetView()
-  Qgs3DMapConfigWidget *w = new Qgs3DMapConfigWidget( map, mMainCanvas, mCanvas, &dlg );
-  QDialogButtonBox *buttons = new QDialogButtonBox( QDialogButtonBox::Ok | QDialogButtonBox::Apply | QDialogButtonBox::Cancel | QDialogButtonBox::Help, &dlg );
+  Qgs3DMapConfigWidget *w = new Qgs3DMapConfigWidget( map, mMainCanvas, mCanvas, dlg );
+  QDialogButtonBox *buttons = new QDialogButtonBox( QDialogButtonBox::Ok | QDialogButtonBox::Apply | QDialogButtonBox::Cancel | QDialogButtonBox::Help, dlg );
```
Hmm if we make the dialog non-modal (which will be a good thing), I think it is maybe more appropriate to have just Close button instead of OK/Cancel which seem more appropriate for modal dialogs... what do you think?

ptitjano replied:
It looks like a good idea. I have updated the PR with this change.

nyalldawson commented:
@ptitjano have you made sure there's no issues if eg the 3d window is closed before the configure dialog? What about if the project is cleared or a new project loaded while the dialog is open? Or layers removed from the project which are referenced in the dialog (e.g. the terrain source)?

I strongly suspect that you'll need to catch these cases and handle them gracefully, and that the current PR will lead to crashes in these scenarios.

ptitjano commented:
> @ptitjano have you made sure there's no issues if eg the 3d window is closed before the configure dialog? What about if the project is cleared or a new project loaded while the dialog is open? Or layers removed from the project which are referenced in the dialog (e.g. the terrain source)?
> 
> I strongly suspect that you'll need to catch these cases and handle them gracefully, and that the current PR will lead to crashes in these scenarios.

Closing the 3d window before the configure dialog does not create any issue. The configure dialog is properly closed.
The terrain source uses `QgsMapLayerComboBox` which properly handles layers deletion
Loading a new project does not create any issue. The dialog and the 3d window are closed.

ptitjano committed:
qgs3dmapcanvaswidget: Use close button for the configuration dialog

With this dialog being non-modal (see previous commit), it is more
appropriate to have just the Close button instead of OK/Cancel.

ptitjano commented:
@wonder-sk Is there anything blocking this MR?

wonder-sk reviewed (approved):
Let's give it a try :-)

ptitjano commented:
@wonder-sk can you merge it please?

nyalldawson commented:
@ptitjano 

I get a crash after this -- open a 3d view for the first time, and change the "terrain" to "dem" -- you don't get a chance to pick the layer and QGIS will crash

ptitjano commented:
> @ptitjano
> 
> I get a crash after this -- open a 3d view for the first time, and change the "terrain" to "dem" -- you don't get a chance to pick the layer and QGIS will crash

Thanks.  That's because, the `OK` button has been removed. I'm gonna create a PR to fix it.

ptitjano commented:
> @ptitjano
> 
> I get a crash after this -- open a 3d view for the first time, and change the "terrain" to "dem" -- you don't get a chance to pick the layer and QGIS will crash

fixed by https://github.com/qgis/QGIS/pull/50028

uclaros commented:
You also shouldn't be allowed to spawn more than one settings dialog per 3d view... When selecting *Configure...* the already open settings dialog should be activated

ptitjano commented:
> You also shouldn't be allowed to spawn more than one settings dialog per 3d view... When selecting _Configure..._ the already open settings dialog should be activated

Indeed. Please see https://github.com/qgis/QGIS/pull/50059
